### PR TITLE
[Qt] prefer CRecipient over std::pair like in walletmodel

### PR DESCRIPTION
**Pull request by Diapolo**
- this updates Qt code to core code, where CRecipient is used in wallet
- also makes things e.g. in paymentserver.cpp much more readable

### CODE REVIEW

**Comment by Diapolo** [comment]
@theuni What does this mean, can you help me?
``error: source directory already configured; run "make distclean" there first``

**Comment by luke-jr** [comment]
"code code"? (commit message)

**Comment by Diapolo** [comment]
@luke-jr Strange... updated the commit message :). Thanks for catching.

**Comment by fanquake** [comment]
@laanwj Do you want to restart the build for this pull?

**Comment by laanwj** [comment]
There's some include file problem in the disable-wallet build:
```
In file included from ./wallet/walletdb.h:10:0,
from ./wallet/wallet.h:20,
from qt/paymentrequestplus.h:11,
from qt/walletmodel.h:8,
from qt/guiutil.cpp:10:
./wallet/db.h:21:20: fatal error: db_cxx.h: No such file or directory
compilation terminated.
```

**Comment by Diapolo** [comment]
@laanwj The problem I see is that ``CRecipient`` is defined in wallet.h and I couldn't get the walletless build to not error out yet -_- (there are quite some fancy dependencies). I'm not convinced ``CRecipient`` fits best in wallet.h anyway.

**Comment by laanwj** [comment]
I think it's the proper place- why would you need a recipient type at all when not concerned with the wallet?

**Comment by sipa** [comment]
The createrawtransaction RPC could use something like CRecipient, even without wallet support.

**Comment by laanwj** [comment]
I guess so, but that'd be completely internal to that call - which is already very simple, and a pure utility method at that. So you'd have two unconnected parts of the source code somehow using the same data type. No, I don't really see that as useful code change. We already have CTxOut. Other recipient metadata is likely application-specific.

**Comment by Diapolo** [comment]
Nice, seems I got it :), woohooo... all builds are passing.

**Comment by laanwj** [comment]
I've looked over this and I don't really think it's an improvement.

As said in my previous post, the CRecipient type is very much bound to our specific current wallet implementation. There's already at least one field in it that is irrelevant to payment requests. Also there is something to be said for keeping the payment server code somewhat more generic.

If the goal is just to neaten the code a bit, I'd suggest to define a local recipient type in `paymentrequestplus.h`, eg a typedef of `std::pair<CScript,CAmount>`.

File: src/qt/paymentserver.cpp [hunk 1]
```diff
@@ -582,22 +582,22 @@ bool PaymentServer::processPaymentRequest(const PaymentRequestPlus& request, Sen
...
         // and no overflow has happened.
-        if (!verifyAmount(sendingTo.second)) {
+        if (!verifyAmount(sendingTo.nAmount)) {
             emit message(tr("Payment request rejected"), tr("Invalid payment request."), CClientUIInterface::MSG_ERROR);
             return false;
         }
 
         // Extract and check amounts
-        CTxOut txOut(sendingTo.second, sendingTo.first);
+        CTxOut txOut(sendingTo.nAmount, sendingTo.scriptPubKey);
```
Comment (Diapolo) [inline src/qt/paymentserver.cpp]: @laanwj Well I think in terms of understanding the code this was an improvement.

Comment (Diapolo) [inline src/qt/paymentserver.cpp]: But not that much, so you would ACK it, right?

Comment (laanwj) [inline src/qt/paymentserver.cpp]: As said, I'm fine with defining a local recipient type here and using that to clarify the code. I just disagree with using the wallet's recipient type.

**Pull request closed without merging**